avatars/animationHelpers.js

```diff
@@ -1007,49 +1029,54 @@ export const _applyAnimation = (avatar, now, moveFactors) => {
             .sub(localVector2.fromArray(v3))
             .add(localVector2.fromArray(v2));
         }
+
+        _handleEndActionTransition(spec);
       };
     } else if (activeAvatar.aimAnimation) {
       return spec => {
         const {
           animationTrackName: k,
           dst,
           // isTop,
+          isArm,
           isPosition,
         } = spec;
 
-        const aimAnimation = (activeAvatar.aimAnimation && aimAnimations[activeAvatar.aimAnimation]);
         _handleDefault(spec);
-        const t2 = (activeAvatar.aimTime / aimMaxTime) % aimAnimation.duration;
-        if (!isPosition) {
-          if (aimAnimation) {
+
+        const aimAnimation = (activeAvatar.aimAnimation && aimAnimations[activeAvatar.aimAnimation]);
+        if (aimAnimation) {
+          const t2 = (activeAvatar.aimTime / aimMaxTime) % aimAnimation.duration;
+          if (!isPosition) {
             const src2 = aimAnimation.interpolants[k];
             const v2 = src2.evaluate(t2);
+            localQuaternion2.fromArray(v2);
 
-            const idleAnimation = _getIdleAnimation('walk');
-            const t3 = 0;
-            const src3 = idleAnimation.interpolants[k];
-            const v3 = src3.evaluate(t3);
+            if (isArm) {
+              dst.slerp(localQuaternion2, Math.min(0.5, activeAvatar.aimTime / 200 * 0.5));
+            } else {
+              localQuaternion2.slerp(dst, 0.5 + moveFactors.idleWalkFactor * 0.5);
+              dst.slerp(localQuaternion2, Math.min(1, activeAvatar.aimTime / 200));
+            }
+          } else {
+            const src2 = aimAnimation.interpolants[k];
+            const v2 = src2.evaluate(t2);
+            localVector2.fromArray(v2);
+            _clearXZ(localVector2, isPosition);
 
-            dst
-              .premultiply(localQuaternion2.fromArray(v3).invert())
-              .premultiply(localQuaternion2.fromArray(v2));
+            if (isArm) {
+              dst.lerp(localVector2, Math.min(0.5, activeAvatar.aimTime / 200 * 0.5));
+            } else {
+              localVector2.lerp(dst, 0.5 + moveFactors.idleWalkFactor * 0.5);
+              dst.lerp(localVector2, Math.min(1, activeAvatar.aimTime / 200));
+            }
           }
-        } else {
-          const src2 = aimAnimation.interpolants[k];
-          const v2 = src2.evaluate(t2);
-
-          const idleAnimation = _getIdleAnimation('walk');
-          const t3 = 0;
-          const src3 = idleAnimation.interpolants[k];
-          const v3 = src3.evaluate(t3);
-
-          dst
-            .sub(localVector2.fromArray(v3))
-            .add(localVector2.fromArray(v2));
         }
+
+        _handleEndActionTransition(spec);
       };
     } else if (activeAvatar.unuseAnimation && activeAvatar.unuseTime >= 0) {
-      return spec => {
+      return (spec, isLastBone) => {
```

Comment:
Hm this isn't a good approach. If something should happen after the last bone then the caller should do that, not the quaternion handler.

Reply:
This might actually affect performance negatively to pass extra arguments to all bones for all frames.

Reply:
Ok, @avaer , is this proper? https://github.com/webaverse/app/pull/2791/commits/dd0ea6f9ef7307133eebae33b92dd4a2cd7c07d9